add psalm types

orklah opened the pull request:
This is a draft PR with the changes I made for #525 to see the rest of the CI and to discuss about some changes

The increase in type coverage is disappointing because a lot of mixed types still comes from outside.

Still, the changes prepare the way for when 
https://github.com/nikic/PHP-Parser/pull/647
and
https://github.com/vimeo/psalm/pull/2612

will be merged.

orklah commented on src/Reflection/ReflectionProperty.php:
```diff
@@ -448,6 +450,8 @@ private function assertClassExist(string $className) : void
      * @throws NotAnObject
      * @throws ObjectNotInstanceOfClass
+     * 
+     * @psalm-assert object $object
```
I think this is a good addition, but it adds an error on psalm because it complains that when this method is called, the variable can only be an object and this is redundant.

I think we should suppress the error and keep the assert.

Ocramius replied:
I think the assertion could be refined further to `@psalm-assert T`, where `T` is the type for this `ReflectionProperty`'s owner, but that's for another time/discussion.

Indeed, psalm assertions are annoying when put in test scope, because they remove obvious/dumb scenarios :D

orklah commented on src/Reflection/StringCast/ReflectionClassConstantStringCast.php:
```diff
@@ -16,13 +19,14 @@ final class ReflectionClassConstantStringCast
     public static function toString(ReflectionClassConstant $constantReflection) : string
     {
         $value = $constantReflection->getValue();
+        assert($value === null || is_scalar($value));
```
A constant can technically return an array. This makes sure that it doesn't happen

orklah commented on src/Reflection/ReflectionClass.php:
```diff
@@ -874,7 +874,7 @@ public function isInterface() : bool
      * defined traits, this will return an empty array.
      *
-     * @return ReflectionClass[]
+     * @return list<ReflectionClass>
```
This is discussed here: #530 It raises an error with Psalm that I think is legitimate and I don't know how to handle.

orklah commented on src/Reflection/ReflectionClass.php:
```diff
@@ -519,7 +519,7 @@ static function (ReflectionClassConstant $constant) : string {
      * with keys as constant names and values as {@see ReflectionClassConstant} objects.
      *
-     * @return ReflectionClassConstant[] indexed by name
+     * @return array<string, ReflectionClassConstant> indexed by name
```
This raise a new error on Psalm too because their internal stubs expects this to be an array<int, ReflectionCl
assConstant>. Not sure who's right on this

Ocramius replied:
I think the stub may need updating here

orklah replied:
https://3v4l.org/t05Dv I took this example directly from php.net. The index of array is numerical.
I'll add the case to #530 as it is roughly the same issue

Ocramius replied:
Yeh, the question is whether it is a list or not: I think it is a list in core too

orklah replied:
> Yeh, the question is whether it is a list or not: I think it is a list in core too

I think you got me wrong. The library returns `array<string, ReflectionClassConstant>` right now but native Object return `array<int, ReflectionClassConstant>` (and more probably a `list<ReflectionClassConstant>` but this is not the issue).

I'll silence those errors in psalm to allow the merge, but I think those are legitimate issues.

Ocramius replied:
Oh, careful! We have adapters and then there is this stuff: the adapters are supposed to be compliant with internal definitions, but these classes don't need to be.

orklah replied:
> Oh, careful! We have adapters and then there is this stuff: the adapters are supposed to be compliant with internal definitions, but these classes don't need to be.

oh, I get it. But Psalm sees that this class implements Reflection and warn about the type of the interface being different than declared

Ocramius commented on src/NodeCompiler/CompileNodeToValue.php:
```diff
@@ -60,7 +60,7 @@ public function __invoke(Node $node, CompilerContext $context)
      * Compile constant expressions
      *
-     * @return bool|mixed|null
+     * @return scalar|scalar[]|null
```
I think we should use `@psalm-` -prefixed annotations for psalm-specific types, to avoid confusing existing tools like @phpstan or @jetbrains

Ocramius commented on src/Reflection/ReflectionClass.php:
```diff
@@ -68,16 +68,16 @@ class ReflectionClass implements Reflection
+    /** @var array<string, ReflectionProperty>|null */
     private $cachedImmediateProperties;
 
     /** @var ReflectionProperty[]|null */
```
What's the type of this array's keys?

orklah replied:
I added it. I couldn't deduce it at first so I skipped it.

Ocramius commented on src/SourceLocator/Ast/FindReflectionsInTree.php:
```diff
@@ -32,9 +32,13 @@ final class FindReflectionsInTree
     private $functionReflectorGetter;
 
+    /**
+     * @param AstConversionStrategy $astConversionStrategy
```
Shouldn't be needed

Ocramius commented on src/SourceLocator/Ast/Locator.php:
```diff
@@ -27,6 +28,10 @@ class Locator
     private $parser;
 
+    /**
+     * @param Parser $parser
```
Shouldn't be needed

orklah commented:
> Overall excellent work, but we _need_ (unavoidable, sorry :-( ) to use `@psalm-`-prefixed annotations for any types that aren't understood by the wider ecosystem, sorry :S

Will do 👍

Ocramius commented:
Read through review comments.

Note that the `list` type was added to psalm recently

orklah commented:
> Note that the `list` type was added to psalm recently

Yeah but PHPStorm won't read it, so I added @psalm- where possible while keeping @return

I just removed the draft tag. I suppressed or fixed the remaining psalm errors

Ocramius commented:
> > Note that the `list` type was added to psalm recently
> 
> Yeah but PHPStorm won't read it, so I added @psalm- where possible while keeping @return
> 
> I just removed the draft tag. I suppressed or fixed the remaining psalm errors

This was to note that the stubs may not yet use `list`, since it is a recent addition ;-)

Ocramius commented:
Thanks @orklah!

orklah commented:
@Ocramius Thanks for merging :)

I must have broke Psalm on CI though. I added a windows path on xml file. It worked fine locally but it breaks the CI. Can you fix it or should I create a PR?

Ocramius commented:
Please do send a separate patch when you have time: I'm not actively working on the package these weeks, besides following contributions
